openedx/openedx-platform
# build: compile/watch sass with new npm scripts

## Description

`paver` commands are deprecated for managing static assets. Starting in Sumac, only `npm run` commands will be supported for managing static assets.

To ease the transition, both `paver` and `npm run` commands will work in Redwood. However, we want to stop using the *implementations* of the `paver` asset commands right now, as they are blocking the Python 3.11 upgrade. This will also make the removal of `paver` commands more straightforward come Sumac.

So, this commit turns these commands/functions:
* paver compile_sass (used by configuration)
* paver watch_sass (used by configuration and devstack)
* pavelib/assets.py:_compile_sass (used by Tutor)

into very thin wrappers around the new `npm run` commands. Each of these paver routines now raise a loud deprecation warning, including a message of the `npm run` command that the operator can switch to.

We expect no impact to site operators or end users.

### Also included in separate commits

* fix: simplify npm run watch-sass so that it is more reliable 
* docs: npm scripts for static assets are no longer experimental :) 

## Supporting Information

* https://github.com/openedx/edx-platform/issues/31895

## Testing

### With Tutor

* Check out this branch
* Install a theme, such as indigo: `tutor plugins install indigo`
* Mount your local edx-platform repo: `tutor mounts add <path to your edx-platform>`
* `tutor dev stop && tutor local stop`
* `tutor config save`
* `tutor images build openedx openedx-dev`
* Test results of prod Sass build
  * `tutor local start -d`
  * `tutor local do settheme indigo`
  * Browse some Django-rendered pages. Good examples are the Studio outline editor, LMS Course Preview (accessible from any edited Studio unit by hitting "Preview"), the Instructor Dashboard, and the LMS XBlock Views within any course.
  * `tutor local do settheme default`
  * Browse some Django-rendered pages
* Test results of dev Sass build
  * `tutor dev start -d`
  * Browse some Django-rendered pages
  * `tutor dev do settheme indigo`
  * Browse some Django-rendered pages
* Test hot Sass recompilation
  * `tutor dev run lms env EDX_PLATFORM_THEME_DIRS=/openedx/themes npm run watch-sass`
  * `touch "$(tutor config printroot)/env/build/openedx/themes/indigo/lms/static/sass/_background.scss"`
  * Confirm that the theme rebuilds
  * NOTE: The Tutor-recommend command `tutor dev run watchthemes` seems to lead to an infinite recompilation loop any time any themed Sass file is changed. I have authored this PR to be technically compatible with `tutor dev run watchthemes`, but I'm not worried about fixing the infinite recompilation bug here. 

### Without Tutor

To test outside of Tutor, I recommend just building static assets however you normally would. For example, I believe `paver update_assets` will do the trick for folks using devstack and configuration. 

## Merge Considerations

I am aiming for this week (Thu March 28 if possible) in order to unblock the Python upgrade, which needs to be complete for the Redwood cut in late April.

Priorities after merging this:

* https://github.com/openedx/edx-platform/pull/34439 (**This will unblock the Python 3.11 upgrade**)
* Update `./manage.py compile_sass`, which is used for site theme compilation, to use `npm run` directly instead of going through `paver`
* https://github.com/kdmccormick/tutor/pull/37
* https://github.com/kdmccormick/tutor/pull/34
